core/services/relay/evm/chain_reader.go

```diff
@@ -219,6 +220,41 @@ func (cr *chainReader) GetLatestValue(ctx context.Context, readName string, conf
 	return nil
 }
 
+func (cr *chainReader) GetLatestValueWithHeadData(ctx context.Context, readName string, confidenceLevel primitives.ConfidenceLevel, params any, returnVal any) (*commontypes.Head, error) {
+	if err := cr.GetLatestValue(ctx, readName, confidenceLevel, params, returnVal); err != nil {
+		return nil, err
+	}
+
+	latestHead, finalizedHead, err := cr.ht.LatestAndFinalizedBlock(ctx)
```

Comment:
This block data won't always represent which block the data was pulled from. GetLatestValue resolves finality(block depth) based on confidence levels, you have to get block number in the same way as the event and method binding implementations.

Reply:
Iiliya updated this code